rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,87 @@
+---
+feature: broken-package-removal
+start-date: 2024-07-14
+author: Jörg "Mic92" Thalheim
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+This RFCs defines under what conditions we remove broken and unmaintained leaf packages.
+The RFC does not have the ambition to define all reasons to remove a package but
+instead focuses on simple rules that can be automated.
+
+# Motivation
+
+[motivation]: #motivation
+
+Broken and unmaintained packages still cost us valuable time.
+Broken packages still have to be evaluated to some extend
+and unmaintained packages might still have to rebuild when one of their dependencies changes.
+Removing unmaintained/unused code hopefully increases the overall package quality and helps us
+to save resources.
+Because it is easy to add packages, it should be also easy to remove them, ideally automated.
+
+# Detailed design
+
+[design]: #detailed-design
+
+## Broken Packages
+
+Packages that are marked as broken unconditionally on all platforms should be removed after one full NixOS release cycle.
+If the package had reverse dependencies, those dependencies are also marked as indirectly broken.
+If they are not functional without the broken package, they should be removed as well.
+
+## Unmaintained packages
+
+Packages with an empty maintainer field that does not have packages depending on it, should
+be removed after 2 months. Ideally we have an automation or semi-automation that creates pull requests for that.
+
+If a NixOS module depends on any of removed package, this module gets removed as well, if it is non-functional without the package.
+In `pkgs/top-level/aliases.nix` we can link to the pull request that removed the package,
+to make it easier for people to recover the nix expression.
+
+# Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+-
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Some maintained packages might still have users that won't be able to use the package,
+but the hope is that the removal of unmaintained package will lead to these people
+stepping up as a maintainer. Recovering from the git history should be easy in that case.
```

Comment:
I'm not sure if this is worth, but similar to NixOS module deprecation warnings, we could have a warning something along the lines of:
"The package X has been unmaintained for Y weeks. Its removal is scheduled Z weeks from now; if you would like to continue using this package, you're encouraged to help maintain it."

Recovering from the Git history is fine, but it's additional work that could be saved by a small warning like this IMO. If no one cares for the warning, then we know the removal doesn't significantly hurt anyone.

Reply:
Very good point. I added this to future work as for now I want to focus on policies more rather than implementation. It might be easier to move this to meta though.

Reply:
The Problem framework is perfect for this.

https://github.com/NixOS/rfcs/pull/127